src/CodeGen/GenerateSyntax.hs

```diff
@@ -13,32 +13,62 @@ module CodeGen.GenerateSyntax
   con <- toConProduct datatypeName fields
-  pure $ DataD [] name [] Nothing [con] [ DerivClause Nothing [ ConT ''Eq, ConT ''Ord, ConT ''Show ] ]
-datatypeForConstructors (LeafType (DatatypeName datatypeName) Anonymous) = do
+  result <- symbolMatchingInstance language name datatypeName
+  pure $ DataD [] name [] Nothing [con] [ DerivClause Nothing [ ConT ''TS.Building, ConT ''Eq, ConT ''Ord, ConT ''Show, ConT ''Generic ] ]:result
+datatypeForConstructors language (LeafType (DatatypeName datatypeName) Anonymous) = do
   let name = toName' Anonymous datatypeName
   con <- toConLeaf Anonymous (DatatypeName datatypeName)
-  pure $ DataD [] name [] Nothing [con] [ DerivClause Nothing [ ConT ''Eq, ConT ''Ord, ConT ''Show ] ]
-datatypeForConstructors (LeafType (DatatypeName datatypeName) named) = do
+  result <- symbolMatchingInstance language name datatypeName
+  pure $ DataD [] name [] Nothing [con] [ DerivClause Nothing [ ConT ''TS.Building, ConT ''Eq, ConT ''Ord, ConT ''Show, ConT ''Generic ] ]:result
+datatypeForConstructors language (LeafType (DatatypeName datatypeName) named) = do
   let name = toName' named datatypeName
   con <- toConLeaf named (DatatypeName datatypeName)
-  pure $ NewtypeD [] name [] Nothing con [ DerivClause Nothing [ ConT ''Eq, ConT ''Ord, ConT ''Show ] ]
+  result <- symbolMatchingInstance language name datatypeName
+  pure $ NewtypeD [] name [] Nothing con [ DerivClause Nothing [ ConT ''TS.Building, ConT ''Eq, ConT ''Ord, ConT ''Show, ConT ''Generic ] ]:result
+
+-- | Create TH-generated SymbolMatching instances for sums, products, leaves
+symbolMatchingInstance :: Ptr TS.Language -> Name -> String -> Q [Dec]
+symbolMatchingInstance language name str = do
+  tsSymbol <- runIO $ withCString str (pure . TS.ts_language_symbol_for_name language)
+  tsSymbolType <- pure . toEnum $ TS.ts_language_symbol_type language tsSymbol
+  [d|instance TS.SymbolMatching $(conT name) where
+      showFailure _ node = "Expected " <> $(litE (stringL (show name))) <> " but got " <> show (TS.fromTSSymbol (nodeSymbol node) :: $(conT (mkName "Grammar.Grammar")))
+      symbolMatch _ node = TS.fromTSSymbol (nodeSymbol node) == $(conE (mkName $ "Grammar." <> TS.symbolToName tsSymbolType str))|]
+
+symbolMatchingInstanceForSums ::  Ptr TS.Language -> Name -> [MkType] -> Q [Dec]
+symbolMatchingInstanceForSums _ name subtypes =
+  [d|instance TS.SymbolMatching $(conT name) where
+      showFailure _ node = "Expected " <> $(litE (stringL (show (map extractn subtypes)))) <> " but got " <> show (TS.fromTSSymbol (nodeSymbol node) :: $(conT (mkName "Grammar.Grammar")))
+      symbolMatch _ = $(foldr1 mkOr (perMkType `map` subtypes)) |]
+  where perMkType (MkType (DatatypeName n) named) = [e|TS.symbolMatch (Proxy :: Proxy $(conT (toName' named n))) |]
+        mkOr lhs rhs = [e| (||) <$> $(lhs) <*> $(rhs) |]
+        extractn (MkType (DatatypeName n) Named) = toCamelCase n
+        extractn (MkType (DatatypeName n) Anonymous) = "Anonymous" <> toCamelCase n
```

Comment:
💡 It might be worth factoring `extractn` out as e.g. `typeName :: MkType -> String`; then `toName'` could actually be replaced with `mkName . typeName` or something like that.

Reply:
Also going to handle as part of `toName` refactor in a separate PR.